include/seastar/core/sstring.hh

```diff
@@ -103,7 +103,19 @@ class basic_sstring {
     char_type* str() noexcept {
         return is_internal() ? u.internal.str : u.external.str;
     }
-
+    static constexpr int compare_chars(const char_type* s1, const char_type* s2, size_t n) {
+        if constexpr (sizeof(char_type) == sizeof(char)) {
+            return std::char_traits<char>::compare(reinterpret_cast<const char*>(s1),
```

Comment:
@michoecho hi Michał, thank you for your review. your arguments make a lot of sense. i changed my implementation to check the signness of `char_type` instead.

Reply:
You missed the point. It's *impossible* to optimize the comparison of `signed char`/`int8_t` strings using `memcmp`, because `signed char` strings require signed comparisons and `memcmp` performs unsigned comparisons, which give different results.

`char` and `unsigned_char` strings are defined by the standard  (through `std::char_traits`) as having unsigned comparison (no matter if `char` is signed or unsigned on the given platform). Therefore `char` and `unsigned char` string comparison can be implemented with `memcmp`.

But `signed char` strings have signed comparison and *can't* be implemented with `memcmp`.

You can't fix this patch. It's premise is wrong.

Reply:
discussed with Michał offline, so the following facts that:

1. the standard library compares two `std::string` s with memcmp()
2. that `char` is signed by default

does not prove that we can compare `signed char` with memcmp(), as the standard requires the `std::string` to be compared as a unsigned string, no matter if char is signed or unsigned.